runtime/allocatable.h

```diff
@@ -0,0 +1,88 @@
+
+// Defines APIs for Fortran runtime library support of code generated
+// to manipulate and query allocatable variables, dummy arguments, & components.
+#ifndef FORTRAN_RUNTIME_ALLOCATABLE_H_
+#define FORTRAN_RUNTIME_ALLOCATABLE_H_
+#include "descriptor.h"
+#include "entry-names.h"
+
+namespace Fortran::runtime {
+extern "C" {
+
+// Initializes the descriptor for an allocatable of intrinsic or derived type.
+// The incoming descriptor is treated as (and can be) uninitialized garbage.
+// Must be called for each allocatable variable as its scope comes into being.
+// The storage for the allocatable's descriptor must have already been
+// allocated to a size sufficient for the rank, corank, and type.
+// A descriptor must be initialized before being used for any purpose,
+// but needs reinitialization in a deallocated state only when there is
+// a change of type, rank, or corank.
+void RTNAME(AllocatableInitIntrinsic)(
+    Descriptor &, TypeCategory, int kind, int rank = 0, int corank = 0);
+void RTNAME(AllocatableInitCharacter)(Descriptor &, SubscriptValue length,
+    int kind = 1, int rank = 0, int corank = 0);
+void RTNAME(AllocatableInitDerived)(
+    Descriptor &, const DerivedType &, int rank = 0, int corank = 0);
+
+// For MOLD= and SOURCE= allocation with no (other) prior initialization.
+void RTNAME(AllocatableInitMold)(Descriptor &, const Descriptor &);
+
+// Explicitly sets the bounds and length type parameters of an initialized
+// deallocated allocatable.  (Not for use with MOLD= or SOURCE= in an
+// ALLOCATE statement; see below).  Length type parameters are indexed in
+// declaration order; i.e., 0 is the first length type parameter in the
+// deepest base type.
+void RTNAME(AllocatableSetBounds)(
+    Descriptor &, int dim, SubscriptValue lower, SubscriptValue upper);
+void RTNAME(AllocatableSetDerivedLength)(
+    Descriptor &, int which, SubscriptValue);
+
+// The upper bound is ignored for the last codimension.
+void RTNAME(AllocatableSetCoBounds)(
+    Descriptor &, int codim, SubscriptValue lower, SubscriptValue upper = 0);
+
+// Allocates an allocatable.  The allocatable descriptor must have been
+// initialized and its bounds and length type parameters set and must be
+// in a deallocated state.  For SOURCE= allocation of an uninitialized
+// descriptor, use AllocatableInitMold() before the first allocation.
+// On failure, if hasStat is true, returns a nonzero error code for
+// STAT= and (if present) fills in errMsg; if hasStat is false, the
+// image is terminated.  On success, leaves errMsg alone and returns zero.
+// Successfully allocated memory is initialized if the allocatable has a
+// derived type, and is always initialized by AllocatableAllocateSource().
+int RTNAME(AllocatableAllocate)(
+    Descriptor &, bool hasStat = false, Descriptor *errMsg = nullptr);
+int RTNAME(AllocatableAllocateSource)(Descriptor &, const Descriptor &source,
+    bool hasStat = false, Descriptor *errMsg = nullptr);
```

Comment:
Would `AllocatableAllocate` and `AllocatableAllocateSource` be responsible to check that coarray allocation on other images succeeded or will it have to be done separably ?

Reply:
That check should be done in these runtime library routines, I think.  Synchronization, too.  Unless you think there's a need for compiler-generated code to interact with those actions; they seem to me to be messy details best left to the library.